File: incre-tests/synduce/unimodal_lists/sum.f
Inductive UList = unil Unit | uelt Int | usplit {UList, Int, Int, UList};
Inductive List = nil Unit | cons {Int, List};

repr = (fix (
  \f: List -> UList -> List. \res: List. \xs: UList.
  match xs with
    unil _ -> res
  | uelt x -> cons {x, res}
  | usplit {x, a, b, y} -> f (cons {a, cons {b, f res y}}) x
  end
)) (nil unit);

is_unimodal =
  let aux_down = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      nil _ -> true
    | cons {h, t} -> and (>= pre h) (f h t)
    end
  ) in let aux_up = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      nil _ -> true
    | cons {h, t} -> if <= pre h then f h t else aux_down h t
    end
  ) in \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} -> aux_up h t
  end;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

target = fix (
  \f: UList -> Compress UList. \xs: UList.
  match xs with
    unil _ -> unil unit
  | uelt x -> uelt x
  | usplit {x, a, b, y} -> usplit {f x, a, b, f y}
  end
);

main = \xs: UList. if is_unimodal (repr xs) then spec (repr (target xs)) else 0;

File: ml-benchmark/synduce/unimodal_lists/sum.ml
type ulist = Unil of unit | Uelt of int | Usplit of ulist * int * int * ulist
type list_ = Nil of unit | Cons of int * list_

let repr =
  let rec repr_f res xs =
    match xs with
    | Unil _ -> res
    | Uelt x -> Cons (x, res)
    | Usplit (x, a, b, y) ->
        let fy = repr_f res y in
        let cb = Cons (b, fy) in
        let ca = Cons (a, cb) in
        repr_f ca x
  in
  fun xs -> repr_f (Nil ()) xs

let is_unimodal =
  let rec aux_down pre xs =
    match xs with
    | Nil _ -> true
    | Cons (h, t) -> (pre >= h) && aux_down h t
  and aux_up pre xs =
    match xs with
    | Nil _ -> true
    | Cons (h, t) ->
        if pre <= h then aux_up h t else aux_down h t
  in
  fun xs ->
    match xs with
    | Nil _ -> true
    | Cons (h, t) -> aux_up h t

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> h + spec t

let rec target xs =
  match xs with
  | Unil _ -> Unil ()
  | Uelt x -> Uelt x
  | Usplit (x, a, b, y) -> Usplit (target x, a, b, target y)

let program xs =
  if is_unimodal (repr xs) then spec (repr (target xs)) else 0